### ISSUE TITLE
Ranged attributes in 2.2.1

### ISSUE THREAD

**Issue by cbf305** [cannatag/ldap3#314]
I was updating my code from LDAP3 1.4.0 to 2.2.1 and I think I found an issue with the ranged search, unless I am just not understanding what it's supposed to do.  With `auto_range=True` on the connection I cannot get ranged searching to function properly or with any consistency against AD.  If I set that to False, it works as it did under 1.4.0.

With it on, no matter what group I query, if it's over 1500, that is all that gets returned.  If I try and specify the next 1500 range value I either get the same values again, or it gives me 3000 entries exactly even if the group has more members.  It's quite bizarre.

With it off, I can specify any range values up to 1500 in total and I get the values I am expecting and in the correct quantity.

Thanks!

**Comment by cannatag** [comment]
Hi, 
can you please show me your code? auto_range works well in my tests: when it's True additional searches are performed until the range is exhausted.

Bye,
Giovanni

**Comment by cbf305** [comment]
The code below works under 1.4.0 with slight mods to work under 2.2.1 with the `auto_range` set to `False`.  If I switch to `True`, and run `conn.search(self.base_dn, '(&(objectclass=group)(samaccountname=mygroup)), attributes=['member'])` it fails to return all values; it limits the results to 1500 every time or goes goofy as I mentioned before with the range specified.

Working code for `auto_range=False`
```
# Get the members
conn.search(base_dn,
				'(&(objectclass=group)(samaccountname={}))'.format(gid),
				attributes=['member'])

# Get the returned attribute names and check if there is a range
attribs = conn.entries[0].entry_attributes
if len(attribs) == 1:
	tmp = conn.entries[0].member.values
else:  # We have a ranged response
	attribs = attribs[1]
	upper_range = attribs.split('-')[1]
	tmp = []
	while upper_range <> '*':
		tmp += conn.entries[0][attribs].values
		lower_range = str(int(upper_range) + 1)
		upper_range = str(int(lower_range) + 1499)
		new_range = 'member;range=%s-%s' % (lower_range, upper_range)
		conn.search(base_dn,
						'(&(objectclass=group)(samaccountname={}))'.format(gid),
						attributes=[new_range])
		attribs = conn.entries[0].entry_attributes[0]
		tmp += conn.entries[0][attribs].values
		upper_range = attribs.split('-')[1]

if uid_only:
	return tuple([uid.split(',')[0].split('=')[1] for uid in tmp])
else:
	return tuple(tmp)
```

Our largest group has 7633 members and the above code will get them all every time.  I'm running this against a 2012 DC with the domain forest level at 2008

**Comment by cannatag** [comment]
Hi, it seems that you're not using the auto_range feature at all. With ``auto_range=True`` your code should be:

```
conn.search(base_dn, '(&(objectclass=group)(samaccountname={}))'.format(gid), attributes=['member'])
```
and you should have the 7633 members in the ``member`` attribute. All the range machinery should be trasparently managed by the ldap3.

Can you give a try?

Thanks,
Giovanni

**Comment by cbf305** [comment]
Yea, sorry, I guess I was explaining form the other direction.  Sorry if that caused confusion.

The following code does NOT work:

```
>>> url = ldap3.Server('ldaps://XXX:636')
>>> pool = ldap3.ServerPool([url, ])
>>> conn = ldap3.Connection(pool, UID, PASSWD, auto_range=True)
>>> conn.search('dc=XXX,dc=XXX', '(&(objectclass=group)(samaccountname=XXX))', attribute=['member'])
```
If I show the returned attributes, I get what I would expect, just `member`
```
>>> conn.entries[0].entry_attributes
[u'member']
```

Testing with our largest group of 7633 members, if I get the len of the members returned, it should be 7633 but it's not:
```
>>> len(conn.entries[0].member.values)
1500
```

**Comment by cannatag** [comment]
Please try this code:

```
from ldap3 import Connection, Server, ServerPool
pool = ServerPool([Server('labldap01.cloudapp.net')])
conn = Connection(pool, 'CN=John Fano,CN=Users,DC=AD2012,DC=LAB', 'Ab123456cdef', auto_range=True)
conn.bind()
conn.search('dc=ad2012,dc=lab', '(&(objectclass=group)(samaccountname=testgrp))', attributes='member')
print(len(conn.entries[0].member.values))
conn.unbind()

```
This should connect to a AD 2012 server in my test lab. The "testgrp" group contains 1997 members and are returned in a single search. 

Does it work for you? 

If it works can you adapt to your environment and test it against your actual server? You should get 7633 in your case

Thanks,
Giovanni

**Comment by cbf305** [comment]
My apologies for the delay in getting back...life :-|

I ran the code you provided and I did get 1997 member back.  I altered it for my system and it was still limited at 1500.  So looks like there are settings in my system that are causing issues here. All the ntdsutil policy values are at the defaults, so I guess I'll have to stick with doing the ranging myself until I can track down why the results are being limited.

Thanks for helping me sort it out :-)

**Comment by cannatag** [comment]
Please check code in dev. I've updated the auto_range code as per #336

**Comment by cannatag** [comment]
Will be released in 2.2.3

**Comment by cannatag** [comment]
Released in ldap3 v. 2.2.3.
